lib/Dialect/FIRRTL/Transforms/IMDeadCodeElim.cpp

```diff
@@ -259,8 +298,63 @@ void IMDeadCodeElimPass::forwardConstantOutputPort(FModuleOp module) {
 void IMDeadCodeElimPass::runOnOperation() {
   LLVM_DEBUG(llvm::dbgs() << "===----- Remove unused ports -----==="
                           << "\n");
-  auto circuit = getOperation();
-  instanceGraph = &getAnalysis<InstanceGraph>();
+  auto circuits = getOperation().getOps<CircuitOp>();
+  if (circuits.empty())
+    return;
+
+  auto circuit = *circuits.begin();
+  InstanceGraph theInstanceGraph(circuit);
+  instanceGraph = &theInstanceGraph;
+
+  mlir::SymbolTable theSymbolTable(circuit);
+  circt::hw::InnerSymbolTableCollection innerSymTables;
+  if (failed(innerSymTables.populateAndVerifyTables(circuit)))
+    return;
+
+  circt::hw::InnerRefNamespace theInnerRefNamespace{theSymbolTable,
+                                                    innerSymTables};
+  symbolTable = &theSymbolTable;
+  innerRefNamespace = &theInnerRefNamespace;
+
+  // Walk attributes and find unknown uses of inner symbols or hierpaths.
+  getOperation().walk([&](Operation *op) {
+    if (isa<FModuleOp>(op)) // Port or module annoations are ok.
+      return;
+
+    if (auto hierPath = dyn_cast<hw::HierPathOp>(op)) {
+      auto namePath = hierPath.getNamepath().getValue();
+      // If the hierpath is public or ill-formed, the verifier should have
+      // caught the error. Conservatively mark the symbol as alive.
+      if (hierPath.isPublic() || namePath.size() <= 1 ||
+          namePath.back().isa<hw::InnerRefAttr>()) {
+        markAlive(hierPath);
+        return;
+      }
+
+      auto instanceAttr = namePath.drop_back().back().cast<hw::InnerRefAttr>();
+      instanceToHierpath[instanceAttr].push_back(hierPath);
+      return;
+    }
+
+    // If there is an unknown use of inner sym or hierpath, just mark all of
+    // them alive.
+    for (NamedAttribute namedAttr : op->getAttrs()) {
+      namedAttr.getValue().walk([&](Attribute subAttr) {
+        if (auto innerRef = dyn_cast<hw::InnerRefAttr>(subAttr)) {
+          if (auto instance = dyn_cast_or_null<firrtl::InstanceOp>(
```

Comment:
why is only the instance being marked alive, what if its an `InnerRef` to a  wire, should we mark the wire also live ?

Reply:
Actually declarations(wire/node/reg) with inner symbols are marked as alive in `markDeclaration` function. I want to keep this PR to delete only dead hierpath(and associated instances) so this PR doesn't handle dead inner symbols other than instances. Ideally IMDCE should inner symbols on declarations too but I think it would require extra work.

Reply:
Makes sense, maybe we can file an issue for a future enhancement? I can imagine the same situation with a declaration other than instance.